Add `chop` function for chopping `iso` Strings and Arrays

dipinhora opened the pull request:
For context, my main use case for this functionality is efficient buffer manipulations (i.e. minimal or no copying). I had hoped to be able to accomplish this without modifying `stdlib` but the pointer manipulations required are only possible in `builtin`.

NOTE: This functionality is not safe until the bugs related to `_alloc` for `String` and `Array` trimming are fixed by #2336.

--------------------

This commit adds a `chop` function to both String and Array
that works on an `iso` and chops it in two with the left half
being a new `iso` String/Array reusing the same pointer
and the right half being a `trim_in_place` version of the original
String/Array. Both are returned as `iso` and mutable and do not
share any memory. The operation does not allocate a new pointer
nor copy elements allowing for efficiency.

dipinhora commented:
Updated to add chopping an already chopped string to the tests per #2340.

SeanTAllen commented:
@dipinhora During sync we discussed ambiguity about what happens if the split point is outside the string. For example, a string or array of size 5 and a split point of 10. Can you add test(s) for that to demonstrate what happens as well as update the doc-strings to cover what the user should expect?

Other than that, this is looking good to merge.

dipinhora commented:
@SeanTAllen Doh! Thanks to whoever pointed out that edge case.

I've added a case to both String and Array chop tests for that and fixed the logic to work correctly also.

SeanTAllen commented:
@dipinhora I __think__ it was @Praetonus but I don't remember.
